Vectorize rngs (Issue 45)

bbbales2 opened the pull request:
#### Submission Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

This picks up from where #578 (now closed) left off. Made a new pull to simplify the code.

This attempts to address: #45 (Vectorize rngs)

I used a tiny amount of C++11. I noticed it's in the makefiles. I can work around that if necessary.

I added a bunch of code to generalize the vectorized tests so it'll be easy to expand this to the other distributions. Check https://github.com/bbbales2/math/blob/cf77cba6221e8cd2db701a5563383623daa103b0/test/unit/math/prim/mat/prob/normal_test.cpp for the use.

```twoParamDistCheckThrowsAllTypes``` and ```chiSquareTestAllTypes``` are the two big entry points. Maybe not named the best but oh well for now I guess.

The new testing functions automatically generate simple sets of parameters that satisfy the necessary constraints (Positive, NonNegative, or None) and validate these with the _rngs and combinations of all the different vector types.

What has to be provided is a callable that wraps the _rng function and a callable that generates quantiles that will be used to validate sets of randomly generated numbers. I'd have to add stuff for 1 and 3 argument functions, and it'd have to change to discrete things as well, but it's a start.

Now that I think about it, it might be better to make it where the parameters are generated by parameter generating callables that are passed in. But this way is simpler.

#### How to Verify:

```
./runTests.py test/unit/math/prim/scal/prob/normal_test.cpp
./runTests.py test/unit/math/prim/mat/prob/normal_test.cpp
```

#### Side Effects:

Hopefully none

#### Documentation:

Docs are inline in Doxygen (though I'm not sure Doxygen actually generates docs from test headers)

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): University of California, Santa Barbara

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

bbbales2 commented:
Jenkins, retest this please

bob-carpenter commented:
@seantalts --- if you want to have a go at this one, you can take over the review.  I'm still trying to unbury myself.

bbbales2 commented:
Hmm, Jenkins seems unhappy with this: http://d1m1s1b1.stat.columbia.edu:8080/job/Math%20Pull%20Request%20-%20Tests%20-%20Unit/734/

But I check that test locally and it works fine (./runTests.py test/unit/math/mix/core/operator_equal_test.cpp)

bbbales2 commented on stan/math/prim/scal/meta/VectorBuilder.hpp:
```diff
@@ -26,19 +26,21 @@ namespace stan {
            typename T4 = double, typename T5 = double, typename T6 = double,
            typename T7 = double>
   class VectorBuilder {
```
I made these changes to VectorBuilder so I could compute the return type from data() more easily (this will be the return type for the vectorized rngs)

bbbales2 commented on stan/math/prim/scal/prob/normal_rng.hpp:
```diff
@@ -1,39 +1,66 @@
-    inline double
-    normal_rng(double mu,
-               double sigma,
+    /**
```
A lot of the docs are missing for the RNGs. I've just been adding stuff. It's a lot of boilerplate generic stuff but it might as well be there in case anyone is curious what these things do at the C++ level.

bbbales2 commented on stan/math/prim/scal/prob/normal_rng.hpp:
```diff
@@ -1,39 +1,66 @@
+     * @throw std::invalid_argument if mu and sigma are different lengths
+     */
+    template <typename T_loc, typename T_scale, class RNG>
+    inline typename VectorBuilder<true, double, T_loc, T_scale>::type
```
The return type is computed from VectorBuilder now!

bbbales2 commented on stan/math/prim/scal/prob/normal_rng.hpp:
```diff
@@ -1,39 +1,66 @@
+     * @throw std::domain_error if mu is infinite or sigma is nonpositive
+     * @throw std::invalid_argument if mu and sigma are different lengths
+     */
+    template <typename T_loc, typename T_scale, class RNG>
```
T_loc and T_scale can be any type that scalar_seq_view knows about

bbbales2 commented on stan/math/prim/scal/prob/normal_rng.hpp:
```diff
@@ -1,39 +1,66 @@
 
+      for (size_t n = 0; n < N; n++) {
+        variate_generator<RNG&, normal_distribution<> >
+          norm_rng(rng, normal_distribution<>(mu_vec[n], sigma_vec[n]));
```
Seems a little cludgy to build a new variate generator for every different parameter, but that's what we do.

bob-carpenter replied:
It's a different `mu_vec[n]` and `sigma_vec[n]`.  Are you suggesting generating a vector `s` of normal(0, 1) draws and then scaling with `mu_vec + sigma_vec .* s`?  It'd even be a one-liner if we could write `normal_rng(N, 0, 1)`.

seantalts replied:
Woops, I misread what was happening here, please ignore this comment!

bbbales2 commented on test/unit/math/prim/mat/prob/normal_test.cpp:
```diff
@@ -0,0 +1,42 @@
+using Eigen::Dynamic;
+using Eigen::Matrix;
+
+struct dist_wrapper {
```
This is a functor that allows me to pass a templated _rng around as a variable. It has to be a functor instead of a function itself for the templating to work (https://stackoverflow.com/a/15651770/3769360).

bbbales2 commented on test/unit/math/prim/mat/prob/normal_test.cpp:
```diff
@@ -0,0 +1,42 @@
+  }
+};
+
+std::vector<double> build_quantiles(int N, double mu, double sigma) {
```
This is just code I copied from the scalar test: https://github.com/stan-dev/math/blob/develop/test/unit/math/prim/scal/prob/normal_test.cpp#L29

It's a callable so I can pass it into the other test functions

bbbales2 commented on test/unit/math/prim/mat/prob/normal_test.cpp:
```diff
@@ -0,0 +1,42 @@
+
+TEST(ProbDistributionsNormal, error_check) {
+  twoParamDistCheckThrowsAllTypes(dist_wrapper{},
+                                  Constraint::None, Constraint::Positive);
```
Read this as "Test the distribution _rng dist_wrapper under the assumption that the first parameter is unconstrained and the second must be positive".

This function tests that dist_wrapper throws domain errors when those constraints are violated and throws errors with infs and nans.

It also tests all combinations of input types (two doubles, one double one std::vector, two std::vectors, VectorXd and std::vector, etc.) and throws errors if the sizes of the inputs are incompatible (if both are non-scalar, they must be the same length).

bbbales2 replied:
I'll add a few comments to the normal_test.cpp files. It'll make life easier on the next person that moseys through.

bob-carpenter replied:
+1 to Sean's guideline.  

Also, are you sure you want to use enums like `Constraint::None`?  Our constraints are all over the place (sum to 1, between 0 and 1, between -1 and 1, simplex, etc.).  Wouldn't it be easier to have functions that validate than enums?  Enums are usually a coding strategy of last resort because they require parallel conditionals everywhere they're used.

bbbales2 replied:
Yes, I agree. I'll look at this.

bbbales2 commented on test/unit/math/prim/mat/prob/normal_test.cpp:
```diff
@@ -0,0 +1,42 @@
+  int M = 10;
+
+  chiSquareTestAllTypes(N, M, dist_wrapper{}, build_quantiles,
+                        Constraint::None, Constraint::Positive);
```
Read this as, "Test samples generated by distribution _rng dist_wrapper with the constraint that the first parameter is unconstrained and the second must be positive against the quantiles generated by the build_quantiles function"

Again it tests all the different combos of input types (scalars and vectors). M is the vector length for any vector arguments.

It uses N sample points for each comparison.

bbbales2 commented on test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
```diff
@@ -0,0 +1,364 @@
+#include <boost/random/mersenne_twister.hpp>
+#include <test/unit/math/prim/scal/prob/util.hpp>
+
+enum class Constraint { None, Positive, NonNegative };
```
I think these constraints cover most of the rngs that work on (-inf, inf) and (0, inf).

bbbales2 commented on test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
```diff
@@ -0,0 +1,364 @@
+ * @param N Number of elements that v will be resized to
+ */
+template <typename T1>
+void generateValues(T1& v, Constraint c, int N) {
```
It seems handwavy to have the test values be so arbitrary (-1, 0, 1, 2, etc). I could make it fancier, but I think this does the job fine.

bbbales2 commented on test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
```diff
@@ -0,0 +1,364 @@
+ */
+template <>
+void generateValues(double& v, Constraint c, int N) {
+  if (c == Constraint::None) {
```
This is my biggest gripe. For the scalars I'm only testing one value. Seems brittle.

bbbales2 replied:
I'm just going to hardcode in a few more sets of numbers and have the tests just repeat for each set.

seantalts replied:
Also, I just took a look at some code of Bob's that does similar stuff for some other work and it might be interesting to see how he approached a similar goal: https://github.com/stan-dev/math/blob/develop/test/unit/math/mix/mat/util/autodiff_tester.hpp#L354

bbbales2 commented on test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
```diff
@@ -0,0 +1,364 @@
+                                          Constraint::NonNegative,
+                                          Constraint::None };
+
+  for (auto it1 = constraints.begin(); it1 != constraints.end(); it1++) {
```
c++11! This tests passing in sets of parameters that will work and sets that won't work.

seantalts replied:
Could you do something like `for (auto it1: constraints) {`?

bbbales2 replied:
Yeah haha, Bob mentioned this too.

bbbales2 commented on test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
```diff
@@ -0,0 +1,364 @@
+
+  generateValues(p1, c1, 3);
+  generateValues(p2, c2, 4);
+  if (stan::length(p1) != 1 && stan::length(p2) != 1)
```
Only do this test if both values aren't scalars

seantalts replied:
Why should it throw if both aren't scalars?

bbbales2 replied:
If mu is a vector of length 4 and sigma is a vector of length 3, then the _rngs throw an error cause they need to be the same length.

seantalts replied:
At this point in the code, how do we know that p1 and p2 have different lengths if their lengths aren't both 1? Not saying this isn't true I just don't see it at first glance.

bbbales2 replied:
It's my high quality return-value-as-a-parameter code :(.

```
generateValues(p1, c1, 3); // Make p1 length 3 with constraint c1 if it's vector-like otherwise just assign a single value
generateValues(p2, c2, 4); // Make p2 length 4 with constraint c2 if it's vector-like otherwise just assign a single value
```

bbbales2 replied:
Ah! I think I can change this though. I know the types that will be returned. I don't have to infer them, so I can make them return values.

bob-carpenter replied:
I didn't understand the return-value-as-a-parameter thing.

bbbales2 commented on test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
```diff
@@ -0,0 +1,364 @@
+    }
+    std::vector<double> quantiles = generate_quantiles(N, p1_vec[m], p2_vec[m]);
+
+    assert_matches_quantiles(samples_to_test, quantiles, 1e-6);
```
This is all about calling this function in the end. I think I should change the name of the test function to mirror this (the chiSquare name -- I guess I just copied and pasted it whatever haha)

seantalts commented:
Lemme look into it - it's also using the new Linux node so I am suspicious.
On Mon, Aug 21, 2017 at 14:00 Ben Bales <notifications@github.com> wrote:

> Hmm, Jenkins seems unhappy with this:
> http://d1m1s1b1.stat.columbia.edu:8080/job/Math%20Pull%20Request%20-%20Tests%20-%20Unit/734/
>
> But I check that test locally and it works fine (./runTests.py
> test/unit/math/mix/core/operator_equal_test.cpp)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/605#issuecomment-323811219>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAxJ7Ckr_nsvzPuqQ_b59lJeW55JqeOrks5sacXQgaJpZM4O8ym4>
> .
>

seantalts commented:
So this passed when run on the mac, but I'm pretty worried that the results are dependent on the compiler and/or OS - it might very slightly point to some kind of memory corruption or otherwise use of undefined behavior. I'm going to try this on a couple more compilers and OSes and see what happens.

seantalts commented:
So this same tests fails on develop, so I don't think it has to do with this pull request at least.

bob-carpenter commented:
Why not use foreach rather than the following if you go to C++11?

for (auto it1 = constraints.begin(); it1 != constraints.end(); it1++)

seantalts commented on test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
```diff
@@ -0,0 +1,364 @@
+}
+
+/*
+ * twoParamCheckTrows feeds the given generate_samples callable various
```
typo

seantalts commented on test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
```diff
@@ -0,0 +1,364 @@
+ * @param c1 Constraint of first parameter
+ * @param c2 Constraint of second parameter
+ */
+template<typename T_param1, typename T_param2, typename T_generate_samples>
```
I like the doc for this!

seantalts commented on test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
```diff
@@ -0,0 +1,364 @@
+template<typename T_param1, typename T_param2, typename T_generate_samples>
+void twoParamDistCheckThrows(T_generate_samples generate_samples, Constraint c1,
+                             Constraint c2) {
+  boost::random::mt19937 rng;
```
What a random RNG name

seantalts commented on test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
```diff
@@ -0,0 +1,364 @@
+ * @param c2 Constraint of second parameter
+ */
+template<typename T_param1, typename T_param2, typename T_generate_samples>
+void twoParamDistCheckThrows(T_generate_samples generate_samples, Constraint c1,
```
There's nothing inherently distribution based about this function, is there? It might be acceptable as a framework for checking two parameter functions in general, which might suggest naming things more broadly and moving to a util directory? Also @bob-carpenter has some related test framework stuff, not sure if there's overlap here yet... maybe he can comment

bbbales2 replied:
Good point. I will look at his stuff more closely.

seantalts commented on test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
```diff
@@ -0,0 +1,364 @@
+}
+
+/*
+ * Call twoParamDistCheckThrows with the given arguments and all combinations of
```
Ooph, thanks for doing this explicit enumeration. I think there's some hacks you can use to iterate through template type lists and blah blah but this is probably better and easier to understand.

seantalts commented on test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
```diff
@@ -0,0 +1,364 @@
+  for (int m = 0; m < M; ++m) {
+    std::vector<double> samples_to_test;
+    for (int n = 0; n < N; ++n) {
+      samples_to_test.push_back(samples_to_test_transpose[n][m]);
```
Why not use an eigen vector for easy transposition?

bbbales2 replied:
Cause I got this working :P

bob-carpenter commented on stan/math/prim/scal/prob/normal_rng.hpp:
```diff
@@ -1,39 +1,66 @@
-    normal_rng(double mu,
-               double sigma,
+    /**
+     * Return a pseudorandom Normal variate for the given location and scale
```
Everything is pseudorandom here.  I only used `_rng` on the suffixes to keep things short and not confuse our users.

bob-carpenter commented on stan/math/prim/scal/prob/normal_rng.hpp:
```diff
@@ -1,39 +1,66 @@
+#include <stan/math/prim/scal/meta/scalar_seq_view.hpp>
+#include <stan/math/prim/scal/meta/max_size.hpp>
+#include <boost/random/normal_distribution.hpp>
+#include <boost/random/variate_generator.hpp>
```
The doc is a little unclear at the edges, much like C++.  There are only one or two things I really hated, like the odd-numbered indents in some class declaration stuff.

The big departure is that we allow references and exceptions.  Google doesn't like either one because they make a huge code base harder to understand from the outside (you don't know what exception a function might throw and having pointers helps distinguish what's going to get passed).

bbbales2 commented:
Thanks for the reviews @seantalts and @bob-carpenter . I'll fix up the issues that you pointed out.

Since there weren't any major complaints, I'd like to spend a little time thinking about how to make this work with 1 or 3 arguments before we close this out. I'll get to that sometime on the weekend probly.

bob-carpenter commented:
It's usually easy to review multiple smaller chunks.  But if you'd rather put everything into one pull request, it shouldn't be too bad for this one.

seantalts commented:
Also, I fixed the bug that was causing your tests to fail; if you merge in develop before the next time you run the tests that failing one should pass now :)

bbbales2 commented:
@seantalts I merge the updated develop into my branch with --no-ff right? Is that the only special flag I need?

bbbales2 commented:
@seantalts If this passes the tests, it's ready for review again.

Changes:
* Tried to update style to match Google's stuff
* Replaced constraint enum with validation callables. See comments in (https://github.com/bbbales2/math/blob/bc88c6f1c0a489c34eb3902ad92ab71527f52c4b/test/unit/math/prim/mat/prob/normal_test.cpp)
* Changed how numbers are generated for tests (see NonRandomGenerator in https://github.com/bbbales2/math/blob/bc88c6f1c0a489c34eb3902ad92ab71527f52c4b/test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp). Hopefully makes things less brittle, but doesn't make errors impossible to track down.
* Tried to make assignments more clear. Didn't succeed totally. setValues still writes to its arguments.
* I think this would work fine for 1 parameter distributions if the supplied quantiles generator and random number generator functors just ignored the 2nd argument (the tests are a little inefficient then, but whatever). I can extend it to 3 in the next pull, which hopefully will cover most of the real valued distributions.

This feels like something that should be broken up. One file for the NonRandomGenerator (if this sticks), one file for the CheckDistThrows, and one file for CheckQuantiles?

bbbales2 commented:
I've looked at the errors. It looks like Travis is unhappy because it timed out. Jenkins is failing cause of cpplint errors with the header includes I thought I fixed.

cpplint is saying: "Found C system header after C++ system header. Should be: normal_rng.h, c system, c++ system, other." when I do

```
#include <string>
#include <boost/...>
#include <stan/...>
```

But it is happy when I do:
```
#include <stan/...>
#include <boost/...>
#include <string>
```

I'm confused :/. What am I missing?

bob-carpenter commented:
Google coding style requires the standard libs to be imported last.   You should probably take a look at their style doc if you haven't.  I find it useful just to resolve a bunch of what seem like otherwise arbitrary decisions.

I assume their reasoning here is that then that the standard library templates can be instantiated by anything introduced earlier.

bbbales2 commented:
Cool beans, I'll swap it around. But Google's style guide actually says the opposite: https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes (that's what was confusing me). Maybe they changed it?

sakrejda commented:
If you run our `cpplint` locally it'll tell you what to do. I think Bob is
right.

On Mon, Aug 28, 2017 at 9:59 AM Ben Bales <notifications@github.com> wrote:

> Cool beans, I'll swap it around. But Google's style guide actually says
> the opposite:
> https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes
> (that's what was confusing me). Maybe they changed it?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/605#issuecomment-325361196>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAfA6VuTDAaimablwLkJALOjb2lVadWLks5scsfAgaJpZM4O8ym4>
> .
>

bbbales2 commented:
Swapped things around. Issue was that cpplint classifies anything included as ```<...>``` a system header, which makes this work, and I guess is correct, but it sure seems like the opposite of the intent of the style guide :D.

And bleh, I think I got the style backwards on all my functions and classes. Stan uses lower_case_things_separated_by_underscores for functions, classes, and member functions, right? I changed my stuff to the Google format after looking at the Google style guide but didn't re-read the Stan guide haha.

seantalts commented:
@bbbales2 is right, here's what the google style guide I found says:

> 1. dir2/foo2.h.
> 2. C system files.
>3. C++ system files.
>4. Other libraries' .h files.
>5. Your project's .h files.

I think the issue is that cpplint classifies boost as a C system header.

Yep, we use what you said (maybe called "snake case?") for those things, which is another place we deviate from the Google style guide - the guide asks for PascalCase for type names.

seantalts commented:
No wait, I think you're right Ben - it seems to think anything matching `<*.hpp>` is C system.

bbbales2 commented:
@seantalts Headers should all be fixed, and I've updated the naming conventions. These comments still apply: https://github.com/stan-dev/math/pull/605#issuecomment-325162593, and if it passes tests it's ready for review!

bob-carpenter commented:
Maybe they changed their style guides but didn't update CppLint?

Waaah!  I don't want to have to go and change all our include orders again.

seantalts commented:
Okay, sorry, let me take another look. And let's tell Jenkins to retest, think the test got killed.

Jenkins, retest this please.

seantalts commented on test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
```diff
@@ -1,54 +1,133 @@
+    // There are only .size() possible non_random_numbers to try
+    for (int j = 0; j < non_random_generator_instance.size(); j++) {
+      v = non_random_generator_instance();
+      if (check(v))
```
What about a paradigm where the callable just returns a number that satisfies the constraint and an orthogonal function that fills some (container | scalar) with value(s) from the generator?

bbbales2 replied:
I thought about this, but got stumped at bad values, but I'm not sure why now.

I guess the simplest thing is pass in a list of good values and bad values for each parameter. That sounds simple now that I say it. I'm not sure why I didn't do it that way. Mistakes were made, sigh. I'll switch it to be like this.

This will simultaneously get rid of the non_random_generator, generate_values, and VectorizeCheck (as a side note, was looking at Google source code today and to my amusement found snake case for member functions: https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/core/framework/op_kernel.h#L686 -- looks like they can't quite stick to TheirStyleRecommendations)

seantalts replied:
unrelated to this PR, are you familiar with Tensorflow? It would be awesome to figure out how to do some benchmarks against the math library for autodiff.

seantalts replied:
Also, did you see my other comment? Github is hiding it by default, calling it outdated...? Pasting it below, from vector_rng_test_helper.hpp:172

Oh man, I think I see what you're saying - you're using the implicit knowledge of how your test parameters were generated here. Can we just make an explicit check? I think that'd be more clear and with less invisible coupling.

To explain a little more - I think the way you're passing lengths to generate_values despite the fact that only some calls will use the lengths, then later checking at runtime the lengths is a potential code smell. I might recommend playing with the API a bit more - maybe encapsulate things like the length inside of generate_values and then check what's generated on the outside. As is, the functions here rely on intimate internal knowledge of each other. And it seems a bit strange the way set_value comes in after to reset the values you just generated. I wouldn't be surprised if it had to be this complicated as you're dealing with something abstracting over a lot of different types (which is good!), but I do wonder if it could be a little simpler?

This framework seems like it's halfway to something like quickcheck - but we might be okay with hardcoding inputs of each type and constraint somewhere and trying all combinations. Go prescribes something more like the latter called table-driven testing that's simpler and maybe more appropriate here. I would find it less confusing than hiding the values you test with inside of a non_random_generator that gets passed into generate_values etc.

seantalts commented on test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
```diff
@@ -69,37 +148,38 @@ void setValue(T1& v, double value) {
-  if (c1 == Constraint::None && c2 == Constraint::NonNegative)
-    return false;
+template<typename T_check, typename T_param>
+bool VectorizeCheck(T_check check, T_param p) {
```
This function is typically called something like [every?](https://clojuredocs.org/clojure.core/every_q), [all](http://hackage.haskell.org/package/base-4.10.0.0/docs/Prelude.html#v:all), or [forall](http://www.scala-lang.org/api/2.12.1/scala/collection/immutable/List.html#forall(p:A=>Boolean):Boolean). I'd re-use an existing name for it (whichever is fine), or maybe try to use [something like this](http://www.cplusplus.com/reference/algorithm/all_of/) in the stdlib, though that's unaware of things like `scalar_seq_view`.

seantalts commented on test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
```diff
@@ -1,54 +1,133 @@
+      double v;
+
+      // There are only .size() possible non_random_numbers to try
+      for (int j = 0; j < non_random_generator_instance.size(); j++) {
```
this makes it seem pretty clear that the concept of a `non_random_generator_instance` is just a list, no?  :P

bbbales2 replied:
Haha, yeah, but it saves its state in a weird way. A bunch of things use it, but they just start from where the last instance left off (this is where it's like the random number generator -- things really don't have a clue what they'll see next).

It's unfortunate those generate_values functions aren't member functions. The way I'm using them (to return either a std::vector or a double) I'd need partial function template specialization. Unless the value is returned through an input argument, bleh.

seantalts commented on test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
```diff
@@ -1,54 +1,133 @@
+    double v;
+
+    // There are only .size() possible non_random_numbers to try
+    for (int j = 0; j < non_random_generator_instance.size(); j++) {
```
I think you can replace this loop with something like `none_of`: http://en.cppreference.com/w/cpp/algorithm/all_any_none_of

bbbales2 commented:
@seantalts  If these tests pass, ready for another check.

Dealing with scalars sometimes and vectors sometimes is still a little messy, but I tried to clean it up some and make the function names verbose enough that they'd catch someone's attention.

I implemented passing in lists of valid/invalid values like you suggested. Looks nice. I think this is better than the last version.

bob-carpenter commented:
>  I think this is better than the last version.

Great news.  Just a heads up that this tends to get easier as you get familiar with the metaprogramming idioms and the Stan idioms.  We try to go very carefully through people's early pull requests, then get them to help reviewing the next person's code (like @seantalts is doing now after his heroic effort on refactoring `operands_and_partials`).

bbbales2 commented on test/unit/math/prim/mat/prob/vector_rng_test_helper.hpp:
```diff
@@ -0,0 +1,372 @@
+    for (int i = 0; i < 5; i++) {
+      std::random_shuffle(bad_p1.begin(), bad_p1.end());
+      std::random_shuffle(good_p2.begin(), good_p2.end());
+      assign_parameter_values(p1, bad_p1);
```
I don't think it matters that I'm filling the thing completely with bad values, though I could add a set_one_value function. I just reused assign_parameter_values to avoid too many weird helper functions.

But shuffling the bad values is just silly. I need to ensure specifically that each one causes a fail. I should just loop through the bad values. It's quite possible I could miss one this way.

seantalts replied:
Yeah, it would be good if our tests to be deterministic when possible :P

bbbales2 commented:
@seantalts I made it so that all the bad values get tested.

Btw, re-Jenkins/Travis, should I be doing git merge --no-ffing to keep my branch up to date manually? Or leave that to the automatic merging stuff? I can start if it'll help Jenkins/Travis.

seantalts commented:
I'd slightly prefer testing with just 1 bad value per parameter rather than an entire array of bad values but I think it's fine this way. Looks good! Not sure if @bob-carpenter wants to go over it too to make sure I did a good job code reviewing.

Re: merging - I think Travis will build just the branch, so if there's something in develop that helps your tests pass on Travis merging in would help them pass. I think Jenkins should be building the merged branch. I don't know why you would need the `--no-ff` in these cases, is that written somewhere?

Also, looks like you have one compiler warning to fix (from the Jenkins build).

bbbales2 commented:
Oh haha. Fixed the sign thing. I think Eigen uses longs for its sizes and std::vectors use size_ts, and this function interacts with both types. I static_cast the sizes to size_t now. Maybe using auto on the index would have been better?

The --no-ff is listed here: https://github.com/stan-dev/stan/wiki/Dev:-Git-Process#52--steps-for-pull-requests so I've just been doing it. If it isn't required then great haha. Don't have to worry about forgetting it then.

seantalts commented:
Actually would you mind just using auto? I think that's cleaner and might not require a comment -with C++11 maybe we finally get some situationally non-leaky abstractions?! And then let's let Bob either take a look or decide not to and then we can merge it :) Thanks for going through this with me!

bbbales2 commented:
Jenkins, retest this please

bbbales2 commented:
Well I committed before I thought through this. auto gets the type set from the initializer, not the size.

How do I cancel the Travis jobs? I found the little red xs for Jenkins but I don't get the option to cancel the Travis ones.

seantalts commented:
Are you signed into travis through github? Should show a cancel build button near the top right of the page you get when you click "Details" below.

bbbales2 commented:
Yeah I signed in with my Github. I don't see an option to kill the job. I pm'ed a screenshot on Discourse (in case I'm somehow missing something)

bob-carpenter commented:
I'm happy to look at things if people want me to, but I don't think we should get into the habit of double-reviewing things.  We're all picky in slightly different (and not necessarily even compatible) ways and I don't want to place too large a burden on the person generating a new feature or bug fix.
